Fixes/options label class and attributes

lilHermit opened the pull request:
Allow FormHelper::input to pass options to ControlLabels (labels around Radio and MultiCheckbox)
    Added `optionsLabel`
Added tests
Fixes any tests broken by this

Only slight functionality change is if class was set via the direct methods (FormHelper::multiCheckbox or FormHelper::radio) and the item is checked then 'selected' gets added. Previously it was skipped

lilHermit committed:
Allow FormHelper::input to pass options to ControlLabels (labels around Radio and MultiCheckbox)
    Added `optionsLabel`
Added tests
Fixes any tests broken by this

Only slight functionality change is if class was set via the direct methods (FormHelper::multiCheckbox or FormHelper::radio)
and the item is checked then 'selected' gets added. Previously it was skipped

lilHermit committed:
Fixed phpcs issues

lilHermit committed:
Added addClass method to StringTemplate with accompanying tests
Switched MultiCheckboxWidget and RadioWidget to use addClass

dereuromark commented on src/View/StringTemplate.php:
```diff
@@ -323,4 +324,65 @@ protected function _formatAttribute($key, $value, $escape = true)
+            return $input;
+        }
+
+        $options = $options + [
```
Are you sure thats the direction you want? Having $options being higher prioritized than the second array?

lilHermit replied:
Yes that way it adds defaults and the options take precedence.

I've changed to

```php 
        $options += [
                'useIndex' => 'class',
                'asString' => false
            ];
```

dereuromark commented on src/View/StringTemplate.php:
```diff
@@ -323,4 +324,65 @@ protected function _formatAttribute($key, $value, $escape = true)
+
+        // Convert and sanitise the inputs
+        if (!is_array($class)) {
+            if (is_string($class) && !empty($class)) {
```
Do we need this? The only values allowed are array or string.

lilHermit replied:
Guards against `explode() expects parameter 2 to be string, object given`

Yes it's programmers error but thought it better to handle gracefully

dereuromark commented on src/View/StringTemplate.php:
```diff
@@ -323,4 +324,65 @@ protected function _formatAttribute($key, $value, $escape = true)
+        $class = array_unique(array_merge($class, $newClass));
+
+        if ($options['returnArray'] === false) {
+            $class = implode(" ", $class);
```
You used `'` everywhere else, would be good to keep it consistent as `' '` here, too.

dereuromark commented on tests/TestCase/View/StringTemplateTest.php:
```diff
@@ -300,4 +300,97 @@ public function testPushPopTemplates()
+
+    public function testAddClassMethod()
+    {
+        // Test new class as null, string, array, false etc
```
If you make separate test methods per "scenario", you will not need the inline comments.

testAddClassMethodNewClass
testAddClassMethodCurrentClass

etc for example

dereuromark commented on src/View/StringTemplate.php:
```diff
@@ -323,4 +324,65 @@ protected function _formatAttribute($key, $value, $escape = true)
+     *
+     * - `useIndex` if you are inputting an array with an 'element' called 'class' and want the same returning
+     *                  you should set this to 'class'
+     * - `returnArray` Setting this to `false` will return a space separated string (default is `true`)
```
I am not sure, but I thought we used the other way around here in core: `'asString'` true/false. Defaulting to false.

ADmad commented:
If the aim was to allowing passing options for labels around Radio and MultiCheckbox why did all widget classes have to be modified? The changes looks pretty invasive to me and are very likely to cause problems with app / plugin with custom/overridden widgets like friendofcake/bootstrap-ui.

Also the patch breaks existing tests.

lilHermit commented:
@ADmad The other widgets were changed because `label` needs to be passed to the Widget and you would get `Label="1"` in the input's if not excluded.

I'm just fixing the two failed tests now.

ADmad commented:
Well you shouldn't be passing unneeded options to other widgets. This change will definitely break custom widgets.

ADmad commented:
Your previous patch #10100 has already caused problem with bootstrap ui plugin FriendsOfCake/bootstrap-ui#180 :)

lilHermit commented:
@ADmad I'm happy to limit the extent `label` is passed to widgets (MultiCheckboxWidget & RadioWidget)

Re: #10100 it was a bug though as `nestedInput` was appearing in input rendering

ADmad commented:
> I'm happy to limit the extent label is passed to widgets (MultiCheckboxWidget & RadioWidget)

Please do.

ADmad commented on src/View/StringTemplate.php:
```diff
@@ -323,4 +324,65 @@ protected function _formatAttribute($key, $value, $escape = true)
+     * @param array $options See above for options
+     * @return array|string
+     */
+    public function addClass($input, $newClass, $options = [])
```
For reference [OptionsAwareTrait](https://github.com/FriendsOfCake/bootstrap-ui/blob/master/src/View/Helper/OptionsAwareTrait.php) contains tried and tested code to add new class to existing list.

inoas replied:
Maybe that should be part of cake core then?

lilHermit committed:
Changes to StringTemplate::addClass method
    Defaults to using 'class' element
    Renamed `returnArray` to `asString`

Separated addClass test methods
Fixed tests after StringTemplate::addClass logic change

markstory commented on tests/TestCase/View/Helper/FormHelperTest.php:
```diff
@@ -8496,4 +8498,266 @@ public function testNestedLabelInput()
+            'optionsLabel' => ['class' => ['custom-class', 'another-class'], 'some-other-attr' => 'blah'],
+            'value' => ['1']
+
+        ]);
```
This is a really long test method. Could it be broken up into smaller tests that are co-located with the method under test?

lilHermit replied:
Yes I thought that too, I'll do so

lilHermit committed:
Changed how `label` is passed around, reduces the impact on many Widgets

lilHermit committed:
StringTemplate::addClass now defaults to using 'class' element

lilHermit committed:
Fix so empty class isn't added to RadioWidget render
    (This fixes the tests that were broken)

lilHermit commented on src/View/Helper/FormHelper.php:
```diff
@@ -1008,7 +1008,8 @@ public function input($fieldName, array $options = [])
             'templates' => [],
-            'templateVars' => []
+            'templateVars' => [],
+            'optionsLabel' => true
```
Because the labels show as default, you can now pass `false` to skip the labels wrapping radio/checkboxes

ADmad replied:
`labelOptions` would be a better name IMO.

lilHermit replied:
;-) I wanted to differentiate from the input label, and relates more to the options.

inoas replied:
`optionsLabel` is even ambigious in a way. It could take an array of key/values where the key is the option value and the value is the option label (for radio/checkboxes) or <option>-tag conntent (for select/multi select).

lilHermit replied:
`nestingLabel` / `nestedLabel` ?!

inoas replied:
you are setting options so it should be `*options` ... and you are setting options for labels... so it should be labelsOptions or labelOptions... imho.

lilHermit replied:
`labelOptions` then? 

(I'll make the change tomorrow and push all the other suggestions at the same time)

markstory replied:
What about `labelAttributes` ?

lilHermit replied:
The problem I have with `labelAttributes` is people might think it relates to the group label rather the one wrapping checkbox/radio. Happy to go with whatever though.

markstory replied:
I'm good with `labelOptions` if you think 'attributes' will be confusing.

lilHermit commented on src/View/Widget/MultiCheckboxWidget.php:
```diff
@@ -152,6 +153,9 @@ protected function _renderInputs($data, $context)
             if (!isset($checkbox['templateVars'])) {
                 $checkbox['templateVars'] = $data['templateVars'];
             }
+            if (!isset($checkbox['label'])) {
```
I was sticking with the same sanity check as `$checkbox['templateVars']` it stops the setting of $checkbox each loop of the foreach. Personally think its a bit pointless, I don't think its saving many cycles ;-)

markstory replied:
Ignore me, I'm a muppet. I was misreading the diff and thought this chunk was connected to the chunk above that defaults `label` to `true`.

lilHermit replied:
After further discussion I see it's there in case `$val` is an array

lilHermit commented on src/View/Widget/MultiCheckboxWidget.php:
```diff
@@ -188,21 +192,31 @@ protected function _renderInput($checkbox, $context)
+                $labelAttrs += $checkbox['label'];
+            }
+
+            if (isset($checkbox['checked']) && $checkbox['checked']) {
```
Agreed, that's now done

inoas commented on tests/TestCase/View/Helper/FormHelperTest.php:
```diff
@@ -8496,4 +8496,266 @@ public function testNestedLabelInput()
+        $result = $this->Form->input('test', [
+            'type' => 'radio',
+            'options' => ['A', 'B'],
+            'optionsLabel' => ['class' => 'custom-class']
```
should be `labelOptions`

inoas commented on tests/TestCase/View/Helper/FormHelperTest.php:
```diff
@@ -8496,4 +8496,266 @@ public function testNestedLabelInput()
+        $result = $this->Form->radio('test', ['A', 'B'], [
+            'label' => [
+                'class' => ['custom-class', 'another-class'],
+                'some-other-attr' => 'blah'
```
`data-*` attributes are more understandable.

inoas commented:
> @ADmad I'm happy to limit the extent label is passed to widgets (MultiCheckboxWidget & RadioWidget)

`labelOptions` should imho apply to all checkbox (single, multiple) and radio controls to be consistent.

lilHermit committed:
Make sure 'checked' is always set and remove the isset

ADmad commented on src/View/Helper/FormHelper.php:
```diff
@@ -1049,7 +1053,7 @@ public function input($fieldName, array $options = [])
         }
 
-        $input = $this->_getInput($fieldName, $options);
+        $input = $this->_getInput($fieldName, $options + ['label' => $optionsLabel]);
```
Why not keep the key `optionsLabel` / `labelOptions` here too?  Changing it to `label` is not intuitive.

lilHermit replied:
I can change it

lilHermit replied:
Once we have a decision on the name

ADmad commented on src/View/Helper/FormHelper.php:
```diff
@@ -1850,7 +1856,7 @@ public function submit($caption = null, array $options = [])
         $input = $this->formatTemplate('inputSubmit', [
             'type' => $type,
-            'attrs' => $this->templater()->formatAttributes($options),
+            'attrs' => $this->templater()->formatAttributes($options, ['label']),
```
Why is this change necessary?  Submit buttons have nothing to do with labels so that should never be passed to `submit()` options.

lilHermit replied:
Because it was failing a test `testFormInputSubmit` due to the default method call in _getInput. I forgot to remove it, apologies!

ADmad commented on src/View/Helper/FormHelper.php:
```diff
@@ -1930,6 +1936,8 @@ public function select($fieldName, $options = [], array $attributes = [])
             return $this->multiCheckbox($fieldName, $options, $attributes);
         }
 
+        unset($attributes['label']);
```
This too is probably not necessary.

lilHermit replied:
That is required to stop `label` getting to `SelectWidget` but still to `MultiCheckWidget`

lilHermit replied:
5 test failures without it as `label="1"` gets into `select`

markstory replied:
So then it is necessary :smile:

lilHermit committed:
Changed test attribute to `data-name` more meaningful
Removed rogue `label` element set

lilHermit committed:
Renamed `optionsLabel` to `labelOptions`

lilHermit committed:
Split large test method

ADmad commented on src/View/Helper/FormHelper.php:
```diff
@@ -1008,7 +1008,8 @@ public function input($fieldName, array $options = [])
             'templates' => [],
-            'templateVars' => []
+            'templateVars' => [],
+            'labelOptions' => true
```
This new option should be documented in the method docblock.

ADmad commented on src/View/Widget/MultiCheckboxWidget.php:
```diff
@@ -112,7 +112,8 @@ public function render(array $data, ContextInterface $context)
             'idPrefix' => null,
-            'templateVars' => []
+            'templateVars' => [],
+            'label' => true
```
Why is this key `label` and not `labelOptions` to maintain consistency? :)

lilHermit replied:
Because I deemed it not important within the widget as within that scope it's label

ADmad replied:
The value of `label` would be the same as value of `labelOptions` passed to FormHelper::input() so better to keep the name same.

lilHermit replied:
It has to remain as `label` in `RadioWidget` as it can be called direct from `FormHelper::radio`. Also you could call `FormHelper::multiCheckbox with `label` and it would work. I think it needs to remain as `label` in the widgets for these direct method functionality too.

lilHermit replied:
```php
$this->Form->multiCheckbox(
    'checkbox1',
    [
        ['text' => 'First Checkbox', 'value' => 1],
        ['text' => 'Second Checkbox', 'value' => 2]
    ],
    [
        'multiple' => 'checkbox',
        'label' => ['class' => ['custom-class', 'another-class'], 'data-name' => 'bob'],
        'value' => ['1']
    ]
);
```
Maybe we should just `label` to the docblock of `multiCheckbox`

ADmad replied:
Since `label` was already used in `FormHelper::radio()` I guess it's okay.

lilHermit replied:
I'll add `label` to docblock for multiCheckbox

lilHermit committed:
Updated docblocks

markstory commented on src/View/StringTemplate.php:
```diff
@@ -323,4 +324,68 @@ protected function _formatAttribute($key, $value, $escape = true)
+     *
+     * - `useIndex` if you are inputting an array with an 'element' other than 'class'.
+     *              Also setting to 'false' will manipulate the whole array (default is 'class')
+     * - `asString` Setting this to `true` will return a space separated string (default is `false`)
```
Do we really need to offer array & string based return values? Can we get away with a single return type?

lilHermit replied:
This method comes from a Utility class I have (slightly modified). Ideally this method would live in HtmlHelper or a Html Utility class like `Cake\Utility\Text` so it can be used by end-user. Preferable a static Utility class then its easier to use in Core. With this in mind I think `asString` would be useful. However I'm happy to remove the option in this context.

dereuromark replied:
I also think we should keep the API here simple, and rather manually `implode(' ', $array)` afterwards, easy enough, right?

lilHermit replied:
alo `implode` is quite tricky ;-)

lilHermit replied:
Ok that's pushed. I also changed `$options` to `$useIndex` string param as we only have one option ;-)

lilHermit committed:
Removed the `asString` option from `StringTemplate::addClass` method
Switched from $options array to string `$useIndex` as we only have one option now

markstory commented on src/View/StringTemplate.php:
```diff
@@ -323,4 +324,58 @@ protected function _formatAttribute($key, $value, $escape = true)
+
+        $class = array_unique(array_merge($class, $newClass));
+
+        if (is_string($useIndex)) {
```
When would useIndex not be a string?

lilHermit replied:
If you are passing in an array without associative key as it supports `false` For example
```php
class MyClass
{
    private $defaultClasses = ['class1', 'class2'];
    private $defaultClassesString = 'class1 class2';

    private function myMethod()
    {
        $classes = $this->template->addClass(
            $this->defaultClasses,
            'new_class',
            false
        );
        // or
        $classes = $this->template->addClass(
            $this->defaultClassesString,
            'new_class',
            false
        );

    }
}
```

markstory replied:
Do we currently have a use case for that?

lilHermit replied:
Currently not but surely it might be useful!?! I don't see a reason not too for one line of code!!

markstory replied:
If we don't need it *now*, I prefer to not have 'extra' code. It adds complexity, requires more testing and more documentation.

lilHermit committed:
Removed `$useIndex` support of false

markstory commented on src/View/Helper/FormHelper.php:
```diff
@@ -1463,7 +1472,9 @@ public function checkbox($fieldName, array $options = [])
      * - `value` - Indicates the value when this radio button is checked.
-     * - `label` - boolean to indicate whether or not labels for widgets should be displayed.
+     * - `label` - Either `false` to disable label around the widget or an array of attributes for
+     *    the label tag. `selected` will be added to any classes e.g. `class => 'myclass'` where widget
```
Missing `'` around 'class' here.

markstory commented on src/View/Helper/FormHelper.php:
```diff
@@ -1970,6 +1983,9 @@ public function select($fieldName, $options = [], array $attributes = [])
      * - `hiddenField` - Set to false to remove the hidden field that ensures a value
      *   is always submitted.
+     * - `label` - Either `false` to disable label around the widget or an array of attributes for
+     *   the label tag. `selected` will be added to any classes e.g. `class => 'myclass'` where
```
Missing `'` around 'class' here.

markstory commented on src/View/Widget/RadioWidget.php:
```diff
@@ -173,6 +173,13 @@ protected function _renderInput($val, $text, $data, $context)
         }
+
+        if (!is_bool($data['label'])) {
+            if (isset($radio['checked']) && $radio['checked']) {
```
These conditions could be combined as there are no elses.

markstory commented on tests/TestCase/View/Helper/FormHelperTest.php:
```diff
@@ -8496,4 +8496,278 @@ public function testNestedLabelInput()
+            ],
+            'labelOptions' => ['class' => ['custom-class', 'another-class'], 'data-name' => 'bob'],
+            'value' => ['1']
+
```
Spare newline.

markstory reviewed (commented):
If you could rebase this, I could get it merged in.

lilHermit committed:
Fixed missing quotes in doc block
Removed spare line
Combined conditions

markstory committed:
Merge branch 'master' into fixes/options-label-class-and-attributes

markstory commented:
Merged into master in b138621cb2d99db0e4ebe7cff666f524473a04ba
